Add release notes for Rust 1.83.0 (#133320)
rust-lang/rust: RELEASES.md

Version 1.83.0 (2024-11-28)
==========================

<a id="1.83.0-Language"></a>

Language
--------

> - [The `non_local_definitions` lint now uses a syntactic heuristic instead of type system based analysis](https://github.com/rust-lang/rust/pull/127117)

Reviewer: The previous implementation was stable (even if hidden only in allow-by-default), what about if we compromise talk about both?
```suggestion
- [The `non_local_definitions` lint was refactor and now warns by default](https://github.com/rust-lang/rust/pull/127117)
```

> - [The `non_local_definitions` lint now uses a syntactic heuristic instead of type system based analysis](https://github.com/rust-lang/rust/pull/127117)

Reviewer 2: Yes, reasonable idea.

```suggestion
- [The `non_local_definitions` lint now fires on less code and warns by default](https://github.com/rust-lang/rust/pull/127117)
```

> - [atomics: allow atomic and non-atomic reads to race](https://github.com/rust-lang/rust/pull/128778)

Reviewer 3: Maybe it's worth highlighting `without UB` in the description? 

```suggestion
- [Allow atomic and non-atomic reads to race without UB](https://github.com/rust-lang/rust/pull/128778)
```

> - [Non-exhaustive structs may be empty](https://github.com/rust-lang/rust/pull/128934/)

Author: > #128934 didn't have a relnotes tracking issue

Because it didn't have the actual relnotes label -- now it does and there's #133353.
(I think the relnotes tool picked it up by its FCP label.)

> - [Check WF of source type's signature on fn pointer cast](https://github.com/rust-lang/rust/pull/129021)

Reviewer 2: Moving it to the compat section would seem to make sense.  A particular encoding of the unsound `type_id` hack might be the main thing affected.

> - [Use equality when resolving type dependent paths](https://github.com/rust-lang/rust/pull/129073)

Reviewer 2: The compat section would be better here.  From the PR:

> It seems that this sample code started compiling in 1.0, stopped compiling in 1.54, started compiling again in 1.63, and will stop compiling again in 1.83.

> - [Stabilize `&mut` (and `*mut`) as well as `&Cell` (and `*const Cell`) in const](https://github.com/rust-lang/rust/pull/129195)

Reviewer 2: This one should go at the top.

- [Disallow implicit coercions from places of type `!`](https://github.com/rust-lang/rust/pull/129392)
- [`const extern` functions can now be defined for other calling conventions.](https://github.com/rust-lang/rust/pull/129753)

> - [Allow creating references to statics in `const` initializers](https://github.com/rust-lang/rust/pull/129759)

Reviewer 2: This one should go second from the top.

- [Stabilize `expr_2021` macro fragment specifier in all editions.](https://github.com/rust-lang/rust/pull/129972)

<a id="1.83.0-Compiler"></a>

Compiler
--------

> - [Implement raw lifetimes and labels (`'r#ident`)](https://github.com/rust-lang/rust/pull/126452/)

Reviewer 2: Definitely lang, and worth mentioning toward the top.

- [Add `x86_64-unknown-hurd-gnu` tier 3 target.](https://github.com/rust-lang/rust/pull/128345)
- [Deprecate unsound `-Csoft-float` flag](https://github.com/rust-lang/rust/pull/129897)
- [Fix a soundness bug where rustc wouldn't detect unconstrained higher-ranked lifetimes in a `dyn Trait`'s associated types that occur due to supertraits.](https://github.com/rust-lang/rust/pull/130367)
- [Add `x86_64-unknown-trusty` tier 3 target.](https://github.com/rust-lang/rust/pull/130453)
- [Add `riscv32-wrs-vxworks` and `riscv64-wrs-vxworks` tier 3 targets.](https://github.com/rust-lang/rust/pull/130549)
- [Add `riscv32{e|em|emc}-unknown-none-elf` tier 3 targets.](https://github.com/rust-lang/rust/pull/130555)
- [Add `arm64e-apple-tvos` tier 3 target.](https://github.com/rust-lang/rust/pull/130614)
- [Add `loongarch64-unknown-linux-ohos` tier 3 target.](https://github.com/rust-lang/rust/pull/130750)

Refer to Rust's [platform support page][platform-support-doc]
for more information on Rust's tiered platform support.

<a id="1.83.0-Libraries"></a>

Libraries
---------

> - [Tracking Issue for `waker_getters`](https://github.com/rust-lang/rust/issues/96992/)

Author: Hmm, this should be removed actually because it's in the API list below.

- [`Option::<Option<_>>::flatten`](https://doc.rust-lang.org/stable/core/option/enum.Option.html#method.flatten)
- [`Option::<Result<_, _>>::transpose`](https://doc.rust-lang.org/stable/core/option/enum.Option.html#method.transpose)
- [`RefCell::into_inner`](https://doc.rust-lang.org/stable/core/cell/struct.RefCell.html#method.into_inner)
- [`Result::as_mut`](https://doc.rust-lang.org/stable/core/result/enum.Result.html#method.as_mut)
- [`Result::<&_, _>::copied`](https://doc.rust-lang.org/stable/core/result/enum.Result.html#method.copied)
- [`Result::<&mut _, _>::copied`](https://doc.rust-lang.org/stable/core/result/enum.Result.html#method.copied-1)
- [`Result::<Option<_>, _>::transpose`](https://doc.rust-lang.org/stable/core/result/enum.Result.html#method.transpose)
- [`UnsafeCell::get_mut`](https://doc.rust-lang.org/stable/core/cell/struct.UnsafeCell.html#method.get_mut)
- [`UnsafeCell::into_inner`](https://doc.rust-lang.org/stable/core/cell/struct.UnsafeCell.html#method.into_inner)
- [`array::from_mut`](https://doc.rust-lang.org/stable/core/array/fn.from_mut.html)
- [`char::encode_utf8`](https://doc.rust-lang.org/stable/core/primitive.char.html#method.encode_utf8)

> - [`$float::classify`](https://doc.rust-lang.org/stable/core/primitive.f64.html#method.classify)
> - [`$float::is_finite`](https://doc.rust-lang.org/stable/core/primitive.f64.html#method.is_finite)
> - [`$float::is_infinite`](https://doc.rust-lang.org/stable/core/primitive.f64.html#method.is_infinite)
> - [`$float::is_nan`](https://doc.rust-lang.org/stable/core/primitive.f64.html#method.is_nan)
> - [`$float::is_normal`](https://doc.rust-lang.org/stable/core/primitive.f64.html#method.is_normal)
> - [`$float::is_sign_negative`](https://doc.rust-lang.org/stable/core/primitive.f64.html#method.is_sign_negative)
> - [`$float::is_sign_positive`](https://doc.rust-lang.org/stable/core/primitive.f64.html#method.is_sign_positive)
> - [`$float::is_subnormal`](https://doc.rust-lang.org/stable/core/primitive.f64.html#method.is_subnormal)
> - [`$float::from_bits`](https://doc.rust-lang.org/stable/core/primitive.f64.html#method.from_bits)
> - [`$float::from_be_bytes`](https://doc.rust-lang.org/stable/core/primitive.f64.html#method.from_be_bytes)
> - [`$float::from_le_bytes`](https://doc.rust-lang.org/stable/core/primitive.f64.html#method.from_le_bytes)
> - [`$float::from_ne_bytes`](https://doc.rust-lang.org/stable/core/primitive.f64.html#method.from_ne_bytes)
> - [`$float::to_bits`](https://doc.rust-lang.org/stable/core/primitive.f64.html#method.to_bits)
> - [`$float::to_be_bytes`](https://doc.rust-lang.org/stable/core/primitive.f64.html#method.to_be_bytes)
> - [`$float::to_le_bytes`](https://doc.rust-lang.org/stable/core/primitive.f64.html#method.to_le_bytes)
> - [`$float::to_ne_bytes`](https://doc.rust-lang.org/stable/core/primitive.f64.html#method.to_ne_bytes)

Reviewer: Those `$float` are quite unusual. I see that we have used not too long ago `{integer}`, let's maybe use the placeholder syntax for floats instead.
```suggestion
- [`{float}::classify`](https://doc.rust-lang.org/stable/core/primitive.f64.html#method.classify)
- [`{float}::is_finite`](https://doc.rust-lang.org/stable/core/primitive.f64.html#method.is_finite)
- [`{float}::is_infinite`](https://doc.rust-lang.org/stable/core/primitive.f64.html#method.is_infinite)
- [`{float}::is_nan`](https://doc.rust-lang.org/stable/core/primitive.f64.html#method.is_nan)
- [`{float}::is_normal`](https://doc.rust-lang.org/stable/core/primitive.f64.html#method.is_normal)
- [`{float}::is_sign_negative`](https://doc.rust-lang.org/stable/core/primitive.f64.html#method.is_sign_negative)
- [`{float}::is_sign_positive`](https://doc.rust-lang.org/stable/core/primitive.f64.html#method.is_sign_positive)
- [`{float}::is_subnormal`](https://doc.rust-lang.org/stable/core/primitive.f64.html#method.is_subnormal)
- [`{float}::from_bits`](https://doc.rust-lang.org/stable/core/primitive.f64.html#method.from_bits)
- [`{float}::from_be_bytes`](https://doc.rust-lang.org/stable/core/primitive.f64.html#method.from_be_bytes)
- [`{float}::from_le_bytes`](https://doc.rust-lang.org/stable/core/primitive.f64.html#method.from_le_bytes)
- [`{float}::from_ne_bytes`](https://doc.rust-lang.org/stable/core/primitive.f64.html#method.from_ne_bytes)
- [`{float}::to_bits`](https://doc.rust-lang.org/stable/core/primitive.f64.html#method.to_bits)
- [`{float}::to_be_bytes`](https://doc.rust-lang.org/stable/core/primitive.f64.html#method.to_be_bytes)
- [`{float}::to_le_bytes`](https://doc.rust-lang.org/stable/core/primitive.f64.html#method.to_le_bytes)
- [`{float}::to_ne_bytes`](https://doc.rust-lang.org/stable/core/primitive.f64.html#method.to_ne_bytes)
```

Author: We used `$integer` *more* recently in 1.82, but that's the only time. I agree that braces are probably better to match rustc.

Reviewer 2: I'd expect braces for a type placeholder and `$` for a syntactic one, e.g. `let $pat = $expr;`.

Author: @traviscross then you agree braces are better here?

Reviewer 2: Yes.
